File: U2/EurekaBank_REST_GoogleMaps/Server/EurekaBank_RestFull_Java_GR01/src/java/ec/edu/monster/services/SucursalService.java
package ec.edu.monster.services;

import ec.edu.monster.dal.SucursalDAO;
import ec.edu.monster.dtos.ActualizarSucursalDTO;
import ec.edu.monster.dtos.CoordenadasDTO;
import ec.edu.monster.dtos.CrearSucursalDTO;
import ec.edu.monster.dtos.RespuestaDTO;
import ec.edu.monster.dtos.SucursalDetalleDTO;
import ec.edu.monster.dtos.SucursalResumenDTO;
import ec.edu.monster.models.Sucursal;
import ec.edu.monster.validators.SucursalValidator;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

/**
 * Servicio de lógica de negocio para sucursales
 * 
 * @author EurekaBank
 */
public class SucursalService {

    private final SucursalDAO sucursalDAO;

    public SucursalService() {
        this.sucursalDAO = new SucursalDAO();
    }

    public RespuestaDTO obtenerTodasLasSucursales() {
        try {
            List<Sucursal> sucursales = sucursalDAO.listarTodas();
            List<SucursalResumenDTO> resumenes = new ArrayList<>();

            for (Sucursal s : sucursales) {
                SucursalResumenDTO dto = new SucursalResumenDTO();
                dto.setCodigo(s.getCodigo());
                dto.setNombre(s.getNombre());
                dto.setCiudad(s.getCiudad());
                dto.setContadorCuentas(s.getContadorCuentas());
                dto.setLatitud(s.getLatitud());
                dto.setLongitud(s.getLongitud());
                dto.setDireccion(s.getDireccion());
                dto.setTieneCoordenadas(s.getLatitud() != null && s.getLongitud() != null);

                String dir = s.getDireccion() != null ? s.getDireccion() : "";
                String ciu = s.getCiudad() != null ? s.getCiudad() : "";
                String completa = dir;
                if (!dir.isEmpty() && !ciu.isEmpty()) {
                    completa += ", " + ciu;
                } else if (dir.isEmpty()) {
                    completa = ciu;
                }
                dto.setDireccionCompleta(completa);

                resumenes.add(dto);
            }

            return new RespuestaDTO(true, "Sucursales obtenidas correctamente", resumenes);
        } catch (SQLException ex) {
            return new RespuestaDTO(false, "Error al obtener sucursales: " + ex.getMessage(), "SRV001");
        }
    }

    public RespuestaDTO obtenerSucursalPorCodigo(int codigo) {
        if (codigo <= 0) {
            return new RespuestaDTO(false, "El código de sucursal debe ser mayor a cero", "VAL001");
        }

        try {
            Sucursal sucursal = sucursalDAO.obtenerPorCodigo(codigo);
            if (!SucursalValidator.existe(sucursal)) {
                return new RespuestaDTO(false, "Sucursal no encontrada", "SUC001");
            }

            SucursalDetalleDTO dto = mappingSucursalDetalleDTO(sucursal);

            return new RespuestaDTO(true, "Sucursal encontrada", dto);
        } catch (SQLException ex) {
            return new RespuestaDTO(false, "Error al obtener sucursal: " + ex.getMessage(), "SRV001");
        }
    }

    public RespuestaDTO crearSucursal(CrearSucursalDTO dto) {
        if (dto == null) {
            return new RespuestaDTO(false, "Los datos de la sucursal son requeridos", "VAL002");
        }

        Sucursal sucursal = new Sucursal();
        sucursal.setNombre(dto.getNombre() != null ? dto.getNombre().trim() : null);
        sucursal.setCiudad(dto.getCiudad() != null ? dto.getCiudad().trim() : null);
        sucursal.setDireccion(dto.getDireccion() != null ? dto.getDireccion().trim() : null);
        sucursal.setLatitud(dto.getLatitud());
        sucursal.setLongitud(dto.getLongitud());
        sucursal.setContadorCuentas(0);

        SucursalValidator.ValidacionResultado validacion = SucursalValidator.validarParaCreacion(sucursal);
        if (!validacion.esValida) {
            return new RespuestaDTO(false, validacion.mensajeError, "VAL003");
        }

        try {
            int nuevoCodigo = sucursalDAO.insertar(sucursal);
            if (nuevoCodigo <= 0) {
                return new RespuestaDTO(false, "Error al crear la sucursal", "SUC004");
            }

            Sucursal sucursalCreada = sucursalDAO.obtenerPorCodigo(nuevoCodigo);
            SucursalDetalleDTO detalleDTO = mappingSucursalDetalleDTO(sucursalCreada);

            return new RespuestaDTO(true, "Sucursal creada correctamente", detalleDTO);
        } catch (SQLException ex) {
            return new RespuestaDTO(false, "Error al crear sucursal: " + ex.getMessage(), "SRV001");
        }
    }

    public RespuestaDTO actualizarSucursal(int codigo, ActualizarSucursalDTO dto) {
        if (codigo <= 0 || dto == null) {
            return new RespuestaDTO(false, "Datos inválidos para actualización", "VAL001");
        }

        try {
            Sucursal sucursalExistente = sucursalDAO.obtenerPorCodigo(codigo);
            if (!SucursalValidator.existe(sucursalExistente)) {
                return new RespuestaDTO(false, "Sucursal no encontrada", "SUC001");
            }

            Sucursal sucursalActualizada = new Sucursal();
            sucursalActualizada.setCodigo(codigo);
            sucursalActualizada
                    .setNombre(dto.getNombre() != null ? dto.getNombre().trim() : sucursalExistente.getNombre());
            sucursalActualizada
                    .setCiudad(dto.getCiudad() != null ? dto.getCiudad().trim() : sucursalExistente.getCiudad());
            sucursalActualizada.setDireccion(
                    dto.getDireccion() != null ? dto.getDireccion().trim() : sucursalExistente.getDireccion());
            sucursalActualizada.setContadorCuentas(sucursalExistente.getContadorCuentas());
            // Lógica de actualización parcial de coordenadas: si viene null, mantenemos el
            // existente.
            sucursalActualizada
                    .setLatitud(dto.getLatitud() != null ? dto.getLatitud() : sucursalExistente.getLatitud());
            sucursalActualizada
                    .setLongitud(dto.getLongitud() != null ? dto.getLongitud() : sucursalExistente.getLongitud());

            SucursalValidator.ValidacionResultado validacion = SucursalValidator
                    .validarParaActualizacion(sucursalActualizada);
            if (!validacion.esValida) {
                return new RespuestaDTO(false, validacion.mensajeError, "VAL003");
            }

            boolean actualizado = sucursalDAO.actualizar(sucursalActualizada);
            if (!actualizado) {
                return new RespuestaDTO(false, "Error al actualizar la sucursal", "SUC005");
            }

            // Refetch para asegurar datos persistidos
            Sucursal sucursalFinal = sucursalDAO.obtenerPorCodigo(codigo);
            SucursalDetalleDTO detalleDTO = mappingSucursalDetalleDTO(sucursalFinal);

            return new RespuestaDTO(true, "Sucursal actualizada correctamente", detalleDTO);

        } catch (SQLException ex) {
            return new RespuestaDTO(false, "Error al actualizar sucursal: " + ex.getMessage(), "SRV001");
        }
    }

    public RespuestaDTO eliminarSucursal(int codigo) {
        if (codigo <= 0) {
            return new RespuestaDTO(false, "El código de sucursal debe ser mayor a cero", "VAL001");
        }

        try {
            Sucursal sucursalExistente = sucursalDAO.obtenerPorCodigo(codigo);
            if (!SucursalValidator.existe(sucursalExistente)) {
                return new RespuestaDTO(false, "Sucursal no encontrada", "SUC001");
            }

            if (!SucursalValidator.puedeSerEliminada(sucursalExistente)) {
                return new RespuestaDTO(false, "No se puede eliminar la sucursal porque tiene "
                        + sucursalExistente.getContadorCuentas() + " cuentas asociadas", "SUC007");
            }

            boolean eliminado = sucursalDAO.eliminar(codigo);
            if (!eliminado) {
                return new RespuestaDTO(false, "Error al eliminar la sucursal", "SUC006");
            }

            return new RespuestaDTO(true, "Sucursal eliminada correctamente", null);
        } catch (SQLException ex) {
            return new RespuestaDTO(false, "Error al eliminar sucursal: " + ex.getMessage(), "SRV001");
        }
    }

    public RespuestaDTO actualizarCoordenadas(int codigo, CoordenadasDTO coordenadas) {
        if (codigo <= 0 || coordenadas == null) {
            return new RespuestaDTO(false, "Código o coordenadas inválidas", "VAL001");
        }

        if (!SucursalValidator.coordenadasValidas(coordenadas.getLatitud(), coordenadas.getLongitud())) {
            return new RespuestaDTO(false, "Coordenadas inválidas", "VAL003");
        }

        try {
            Sucursal sucursalExistente = sucursalDAO.obtenerPorCodigo(codigo);
            if (!SucursalValidator.existe(sucursalExistente)) {
                return new RespuestaDTO(false, "Sucursal no encontrada", "SUC001");
            }

            boolean actualizado = sucursalDAO.actualizarCoordenadas(codigo, coordenadas.getLatitud(),
                    coordenadas.getLongitud());
            if (!actualizado) {
                return new RespuestaDTO(false, "Error al actualizar coordenadas", "SUC002");
            }

            return new RespuestaDTO(true, "Coordenadas actualizadas correctamente", null);
        } catch (SQLException ex) {
            return new RespuestaDTO(false, "Error interno: " + ex.getMessage(), "SRV001");
        }
    }

    private SucursalDetalleDTO mappingSucursalDetalleDTO(Sucursal s) {
        SucursalDetalleDTO dto = new SucursalDetalleDTO();
        dto.setCodigo(s.getCodigo());
        dto.setNombre(s.getNombre());
        dto.setCiudad(s.getCiudad());
        dto.setDireccion(s.getDireccion());
        dto.setContadorCuentas(s.getContadorCuentas());
        dto.setLatitud(s.getLatitud());
        dto.setLongitud(s.getLongitud());
        dto.setTieneCoordenadas(s.getLatitud() != null && s.getLongitud() != null);

        String dir = s.getDireccion() != null ? s.getDireccion() : "";
        String ciu = s.getCiudad() != null ? s.getCiudad() : "";
        String completa = dir;
        if (!dir.isEmpty() && !ciu.isEmpty()) {
            completa += ", " + ciu;
        } else if (dir.isEmpty()) {
            completa = ciu;
        }
        dto.setDireccionCompleta(completa);

        return dto;
    }
}

File: U2/EurekaBank_REST_GoogleMaps/EurekaBank_RestFull_DotNet_GR01/EurekaBank_RestFull_DotNet_GR01/services/SucursalService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using EurekaBank_RestFull_DotNet_GR01.DAL;
using EurekaBank_RestFull_DotNet_GR01.Models;
using EurekaBank_RestFull_DotNet_GR01.Models.DTOs;
using EurekaBank_RestFull_DotNet_GR01.Validators;

namespace EurekaBank_RestFull_DotNet_GR01.Services
{
    /// <summary>
    /// Servicio de lógica de negocio para sucursales
    /// </summary>
    public class SucursalService
    {
        private readonly SucursalDAO sucursalDAO;

        public SucursalService()
        {
            sucursalDAO = new SucursalDAO();
        }

        /// <summary>
        /// Obtiene todas las sucursales como DTO de resumen
        /// </summary>
        public RespuestaDTO ObtenerTodasLasSucursales()
        {
            try
            {
                var sucursales = sucursalDAO.ListarTodas();
                
                var sucursalesResumen = sucursales.Select(s => new SucursalResumenDTO
                {
                    Codigo = s.Codigo,
                    Nombre = s.Nombre,
                    Ciudad = s.Ciudad,
                    ContadorCuentas = s.ContadorCuentas,
                    TieneCoordenadas = s.TieneCoordenadas
                }).ToList();

                return new RespuestaDTO
                {
                    Exitoso = true,
                    Mensaje = "Sucursales obtenidas correctamente",
                    Datos = sucursalesResumen
                };
            }
            catch (Exception ex)
            {
                return new RespuestaDTO
                {
                    Exitoso = false,
                    Mensaje = $"Error al obtener sucursales: {ex.Message}",
                    CodigoError = "SRV001"
                };
            }
        }

        /// <summary>
        /// Obtiene una sucursal por código
        /// </summary>
        public RespuestaDTO ObtenerSucursalPorCodigo(int codigo)
        {
            try
            {
                if (codigo <= 0)
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "El código de sucursal debe ser mayor a cero",
                        CodigoError = "VAL001"
                    };
                }

                var sucursal = sucursalDAO.ObtenerPorCodigo(codigo);
                
                if (!SucursalValidator.Existe(sucursal))
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "Sucursal no encontrada",
                        CodigoError = "SUC001"
                    };
                }

                var sucursalDetalle = new SucursalDetalleDTO
                {
                    Codigo = sucursal.Codigo,
                    Nombre = sucursal.Nombre,
                    Ciudad = sucursal.Ciudad,
                    Direccion = sucursal.Direccion,
                    ContadorCuentas = sucursal.ContadorCuentas,
                    Latitud = sucursal.Latitud,
                    Longitud = sucursal.Longitud,
                    TieneCoordenadas = sucursal.TieneCoordenadas
                };

                return new RespuestaDTO
                {
                    Exitoso = true,
                    Mensaje = "Sucursal encontrada",
                    Datos = sucursalDetalle
                };
            }
            catch (Exception ex)
            {
                return new RespuestaDTO
                {
                    Exitoso = false,
                    Mensaje = $"Error al obtener sucursal: {ex.Message}",
                    CodigoError = "SRV001"
                };
            }
        }

        /// <summary>
        /// Crea una nueva sucursal
        /// </summary>
        public RespuestaDTO CrearSucursal(CrearSucursalDTO sucursalDTO)
        {
            try
            {
                if (sucursalDTO == null)
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "Los datos de la sucursal son requeridos",
                        CodigoError = "VAL002"
                    };
                }

                // Crear entidad Sucursal (código se auto-genera)
                var sucursal = new Sucursal
                {
                    Codigo = 0, // Auto-generado en BD
                    Nombre = sucursalDTO.Nombre?.Trim(),
                    Ciudad = sucursalDTO.Ciudad?.Trim(),
                    Direccion = sucursalDTO.Direccion?.Trim(),
                    ContadorCuentas = 0,
                    Latitud = sucursalDTO.Latitud,
                    Longitud = sucursalDTO.Longitud
                };

                // Validar
                var (esValida, mensajeError) = SucursalValidator.ValidarParaCreacion(sucursal);
                if (!esValida)
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = mensajeError,
                        CodigoError = "VAL003"
                    };
                }

                // Insertar y obtener código generado
                int nuevoCodigo = sucursalDAO.Insertar(sucursal);
                if (nuevoCodigo <= 0)
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "Error al crear la sucursal",
                        CodigoError = "SUC004"
                    };
                }

                // Obtener sucursal creada para respuesta completa
                var sucursalCreada = sucursalDAO.ObtenerPorCodigo(nuevoCodigo);

                // Respuesta con DTO
                var sucursalDetalle = new SucursalDetalleDTO
                {
                    Codigo = sucursalCreada.Codigo,
                    Nombre = sucursalCreada.Nombre,
                    Ciudad = sucursalCreada.Ciudad,
                    Direccion = sucursalCreada.Direccion,
                    ContadorCuentas = sucursalCreada.ContadorCuentas,
                    Latitud = sucursalCreada.Latitud,
                    Longitud = sucursalCreada.Longitud,
                    TieneCoordenadas = sucursalCreada.TieneCoordenadas
                };

                return new RespuestaDTO
                {
                    Exitoso = true,
                    Mensaje = "Sucursal creada correctamente",
                    Datos = sucursalDetalle
                };
            }
            catch (Exception ex)
            {
                return new RespuestaDTO
                {
                    Exitoso = false,
                    Mensaje = $"Error al crear sucursal: {ex.Message}",
                    CodigoError = "SRV001"
                };
            }
        }

        /// <summary>
        /// Actualiza una sucursal existente
        /// </summary>
        public RespuestaDTO ActualizarSucursal(int codigo, ActualizarSucursalDTO sucursalDTO)
        {
            try
            {
                if (codigo <= 0 || sucursalDTO == null)
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "El código debe ser mayor a cero y los datos de la sucursal son requeridos",
                        CodigoError = "VAL001"
                    };
                }

                // Verificar que existe
                var sucursalExistente = sucursalDAO.ObtenerPorCodigo(codigo);
                if (!SucursalValidator.Existe(sucursalExistente))
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "Sucursal no encontrada",
                        CodigoError = "SUC001"
                    };
                }

                // Crear entidad actualizada
                var sucursalActualizada = new Sucursal
                {
                    Codigo = codigo,
                    Nombre = sucursalDTO.Nombre?.Trim(),
                    Ciudad = sucursalDTO.Ciudad?.Trim(),
                    Direccion = sucursalDTO.Direccion?.Trim(),
                    ContadorCuentas = sucursalExistente.ContadorCuentas,
                    Latitud = sucursalDTO.Latitud,
                    Longitud = sucursalDTO.Longitud
                };

                // Validar
                var (esValida, mensajeError) = SucursalValidator.ValidarParaActualizacion(sucursalActualizada);
                if (!esValida)
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = mensajeError,
                        CodigoError = "VAL003"
                    };
                }

                // Actualizar
                bool actualizado = sucursalDAO.Actualizar(sucursalActualizada);
                if (!actualizado)
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "Error al actualizar la sucursal",
                        CodigoError = "SUC005"
                    };
                }

                // Respuesta con DTO
                var sucursalDetalle = new SucursalDetalleDTO
                {
                    Codigo = sucursalActualizada.Codigo,
                    Nombre = sucursalActualizada.Nombre,
                    Ciudad = sucursalActualizada.Ciudad,
                    Direccion = sucursalActualizada.Direccion,
                    ContadorCuentas = sucursalActualizada.ContadorCuentas,
                    Latitud = sucursalActualizada.Latitud,
                    Longitud = sucursalActualizada.Longitud,
                    TieneCoordenadas = sucursalActualizada.TieneCoordenadas
                };

                return new RespuestaDTO
                {
                    Exitoso = true,
                    Mensaje = "Sucursal actualizada correctamente",
                    Datos = sucursalDetalle
                };
            }
            catch (Exception ex)
            {
                return new RespuestaDTO
                {
                    Exitoso = false,
                    Mensaje = $"Error al actualizar sucursal: {ex.Message}",
                    CodigoError = "SRV001"
                };
            }
        }

        /// <summary>
        /// Elimina una sucursal
        /// </summary>
        public RespuestaDTO EliminarSucursal(int codigo)
        {
            try
            {
                if (codigo <= 0)
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "El código de sucursal debe ser mayor a cero",
                        CodigoError = "VAL001"
                    };
                }

                // Verificar que existe
                var sucursalExistente = sucursalDAO.ObtenerPorCodigo(codigo);
                if (!SucursalValidator.Existe(sucursalExistente))
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "Sucursal no encontrada",
                        CodigoError = "SUC001"
                    };
                }

                // Validar que puede ser eliminada
                if (!SucursalValidator.PuedeSerEliminada(sucursalExistente))
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = $"No se puede eliminar la sucursal porque tiene {sucursalExistente.ContadorCuentas} cuentas asociadas",
                        CodigoError = "SUC007"
                    };
                }

                // Eliminar
                bool eliminado = sucursalDAO.Eliminar(codigo);
                if (!eliminado)
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "Error al eliminar la sucursal",
                        CodigoError = "SUC006"
                    };
                }

                return new RespuestaDTO
                {
                    Exitoso = true,
                    Mensaje = "Sucursal eliminada correctamente"
                };
            }
            catch (Exception ex)
            {
                return new RespuestaDTO
                {
                    Exitoso = false,
                    Mensaje = $"Error al eliminar sucursal: {ex.Message}",
                    CodigoError = "SRV001"
                };
            }
        }

        /// <summary>
        /// Actualiza coordenadas de una sucursal
        /// </summary>
        public RespuestaDTO ActualizarCoordenadas(int codigo, CoordenadasDTO coordenadas)
        {
            try
            {
                if (codigo <= 0 || coordenadas == null)
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "El código debe ser mayor a cero y las coordenadas son requeridos",
                        CodigoError = "VAL001"
                    };
                }

                // Verificar que existe
                var sucursalExistente = sucursalDAO.ObtenerPorCodigo(codigo);
                if (!SucursalValidator.Existe(sucursalExistente))
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "Sucursal no encontrada",
                        CodigoError = "SUC001"
                    };
                }

                // Validar coordenadas
                if (!SucursalValidator.CoordenadasValidas(coordenadas.Latitud, coordenadas.Longitud))
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "Las coordenadas proporcionadas no son válidas",
                        CodigoError = "VAL003"
                    };
                }

                // Actualizar coordenadas
                bool actualizado = sucursalDAO.ActualizarCoordenadas(codigo, coordenadas.Latitud, coordenadas.Longitud);
                if (!actualizado)
                {
                    return new RespuestaDTO
                    {
                        Exitoso = false,
                        Mensaje = "Error al actualizar las coordenadas",
                        CodigoError = "SUC002"
                    };
                }

                return new RespuestaDTO
                {
                    Exitoso = true,
                    Mensaje = "Coordenadas actualizadas correctamente"
                };
            }
            catch (Exception ex)
            {
                return new RespuestaDTO
                {
                    Exitoso = false,
                    Mensaje = $"Error al actualizar coordenadas: {ex.Message}",
                    CodigoError = "SRV001"
                };
            }
        }
    }
}